modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -44,86 +44,205 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+   * Server supposed to return them in proper order
+   */
+  def listSchemas(
+    vendor: String,
+    name: String,
+    model: Int
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): F[Either[ResolutionError, SchemaList]] =
+    listSchemasResult(vendor, name, model).map(_.map(_.value))
+
   /** Get list of full self-describing schemas available on Iglu Server for particular vendor/name pair */
   def fetchSchemas(
     vendor: Vendor,
     name: Name,
     model: Model
-  )(implicit F: Monad[F], L: RegistryLookup[F], C: Clock[F]) =
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): EitherT[F, ResolutionError, List[SelfDescribingSchema[Json]]] =
     for {
       list <- EitherT(listSchemas(vendor, name, model))
       result <- list.schemas.traverse { key =>
         EitherT(lookupSchema(key)).map(json => SelfDescribingSchema(SchemaMap(key), json))
       }
     } yield result
 
-  private def getFromCache(
+  private def getSchemaFromCache(
     schemaKey: SchemaKey
-  )(implicit F: Monad[F], C: Clock[F]): F[Option[SchemaLookup]] =
+  )(implicit F: Monad[F], C: Clock[F]): F[Option[ResolverCache.TimestampedItem[SchemaLookup]]] =
     cache match {
-      case Some(c) => c.getSchema(schemaKey)
+      case Some(c) => c.getTimestampedSchema(schemaKey)
       case None    => Monad[F].pure(None)
     }
+
+  private def getSchemaListFromCache(
+    vendor: String,
+    name: String,
+    model: Int
+  )(implicit
+    F: Monad[F],
+    C: Clock[F]
+  ): F[Option[ResolverCache.TimestampedItem[ListLookup]]] =
+    cache match {
+      case Some(c) => c.getTimestampedSchemaList(vendor, name, model)
+      case None    => Monad[F].pure(None)
+    }
+
 }
 
 /** Companion object. Lets us create a Resolver from a Json */
 object Resolver {
 
+  type SchemaListKey          = (String, String, Int) // vendor, name, model
+  type SchemaLookupResult     = ResolverResult[SchemaKey, Json]
+  type SchemaListLookupResult = ResolverResult[SchemaListKey, SchemaList]
+
+  /** The result of doing a lookup with the resolver, carrying information on whether the cache was used */
+  sealed trait ResolverResult[+K, +A] {
+    def value: A
+  }
+
+  object ResolverResult {
+
+    /**
+     * The result of a lookup when the resolver is configured to use a cache
+     *
+     * The timestamped value is helpful when the client code needs to perform an expensive
+     * calculation derived from the looked-up value. If the timestamp has not changed since a
+     * previous call, then the value is guaranteed to be the same as before, and the client code
+     * does not need to re-run the expensive calculation.
+     *
+     * @param value     the looked-up value
+     * @param timestamp epoch time in seconds of when the value was last cached by the resolver
+     */
+    case class Cached[K, A](key: K, value: A, timestamp: TTL) extends ResolverResult[K, A]
```

Comment:
Is it more like `StorageTime` instead of `TTL`?

Reply:
Agreed. Have fixed.